packages/discovery-provider/plugins/pedalboard/apps/solana-relay/src/routes/listen/assertListenRateLimiter.test.ts

```diff
@@ -0,0 +1,227 @@
+import { beforeEach, describe, expect, it } from "vitest";
+import { LISTENS_RATE_LIMIT_IP_PREFIX, LISTENS_RATE_LIMIT_TRACK_PREFIX, config } from "../../config";
+import { getRedisConnection } from "../../redis";
+import { listenRouteRateLimiter, listensIpRateLimiter, listensIpTrackRateLimiter } from "./listen";
+
+beforeEach(async () => {
+    // run with npm run test with audius-compose up
+    config.redisUrl = "redis://0.0.0.0:6379/00"
+    const redis = await getRedisConnection()
+    let cursor = '0';
+    do {
+        for await (const key of redis.scanIterator()) {
+            if (key.startsWith(LISTENS_RATE_LIMIT_IP_PREFIX) || key.startsWith(LISTENS_RATE_LIMIT_TRACK_PREFIX)) {
+                await redis.del(key)
+            }
+        }
+    } while (cursor !== '0')
+})
+
+describe('Listens Route Rate Limiter', function () {
+    it('limits independent tracks and users by ip hourly', async function () {
+        const track1 = "track1"
+        const ipOne = "1.2.3.4"
+
+        const track2 = "track2"
+        const ipTwo = "2.3.4.5"
+
+        listensIpRateLimiter.setLimits({ hourlyLimit: 10 })
+        listensIpTrackRateLimiter.setLimits({ hourlyLimit: 40 })
+
+        // ten listens for user one with track one
+        await listenRouteRateLimiter({ ip: ipOne, trackId: track1 })
```

Comment:
```
for (let i = 0; i < 9; i++) {
    await listenRouteRateLimiter({ ip: ipOne, trackId: track1 });
}
```
so the reader doesn't have to count the lines

Reply:
ah good call